middleware/src/decoder/mod.rs

```diff
@@ -0,0 +1,381 @@
+// Copyright 2024 The Drasi Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use base64::Engine;
+use drasi_core::{
+    interface::{
+        ElementIndex, MiddlewareError, MiddlewareSetupError, SourceMiddleware,
+        SourceMiddlewareFactory,
+    },
+    models::{Element, ElementValue, SourceChange, SourceMiddlewareConfig},
+};
+use serde::Deserialize;
+use serde_json::Value;
+
+#[cfg(test)]
+mod tests;
+
+#[derive(Debug, Clone, Deserialize, PartialEq, Default)]
+#[serde(rename_all = "lowercase")]
+pub enum ErrorHandling {
+    #[default]
+    Skip,
+    Fail,
+}
+
+#[derive(Debug, Clone, Deserialize, PartialEq)]
+#[serde(rename_all = "snake_case")]
+pub enum EncodingType {
+    Base64,
+    Base64url,
+    Hex,
+    Url,
+    JsonEscape,
+}
+
+pub struct Decoder {
+    name: String,
+    encoding_type: EncodingType,
+    target_property: String,
+    strip_quotes: bool,
+    parse_json: bool,
+    flatten: bool,
```

Comment:
I think we should move the JSON parsing and flattening to their own middlewares each.